src/IO/IO/Phnx.IO/PipeStream.cs

```diff
@@ -0,0 +1,268 @@
+        {
+            throw new NotSupportedException($"Cannot seek in a {nameof(PipeStream)}");
+        }
+
+        /// <summary>
+        /// Setting the length is not supported in a <see cref="PipeStream"/>
+        /// </summary>
+        /// <exception cref="NotSupportedException">Cannot set the length of a <see cref="PipeStream"/></exception>
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException($"Cannot set the length of a {nameof(PipeStream)}");
+        }
+
+        /// <summary>
+        /// Writes a sequence of bytes to the current stream
+        /// </summary>
+        /// <param name="buffer">An array of bytes. This method copies <paramref name="count"/> bytes from buffer to the current stream</param>
+        /// <param name="offset">The zero-based byte offset in <paramref name="buffer"/> at which to begin copying bytes to the current stream</param>
+        /// <param name="count">The number of bytes to be written to the current stream</param>
+        /// <exception cref="ArgumentException">The sum of offset and count is greater than the buffer length</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is <see langword="null"/></exception>
+        /// <exception cref="ArgumentLessThanZeroException"><paramref name="count"/> or <paramref name="offset"/> is less than zero</exception>
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+            if (offset < 0)
+            {
+                throw new ArgumentLessThanZeroException(nameof(offset));
+            }
+            if (count < 0)
+            {
+                throw new ArgumentLessThanZeroException(nameof(count));
+            }
+            if (offset + count > buffer.Length)
+            {
+                throw new ArgumentException($"{nameof(offset)} and {nameof(count)} combined were greater than the number of elements in {nameof(buffer)}");
+            }
+
+            lock (data)
+            {
+                for (int writeCount = 0; writeCount < count && offset + writeCount < buffer.Length; ++writeCount)
+                {
+                    var valueToWrite = buffer[offset + writeCount];
+
+                    data.Enqueue(valueToWrite);
+                }
+            }
+
+            DataWritten?.Invoke(this);
+        }
+
+        /// <summary>
+        /// Releases the unmanaged resources used by the <see cref="PipeStream"/> and optionally releases the managed resources
+        /// </summary>
+        /// <param name="disposing"><see langword="true"/> to release both managed and unmanaged resources; <see langword="false"/> to release only unmanaged resources</param>
+        protected override void Dispose(bool disposing)
+        {
+            // TODO Add the idea of being disposed
```

Comment:
TODO remains - need to implement Dispose pattern, or remove it. Copy the behaviour in .NET's `MemoryStream`

Reply:
`MemoryStream` does not override `Dispose`, this method should be removed